### ISSUE TITLE
Make `TestPowApprox` and `TestPow` table-driven and add more test cases

### ISSUE THREAD

**Issue by p0mvn** [osmosis-labs/osmosis#1756]
Commenting to create an issue for post v10 work (not in this PR):

Make `TestPowApprox` and `TestPow` table-driven and add more test cases

_Originally posted by @p0mvn in https://github.com/osmosis-labs/osmosis/pull/1755#discussion_r894791047_

**Comment by catShaark** [comment]
@p0mvn, I think we should randomized the test, may be have a python file generating all the random test values to test against. We should treat complex math functions like this since they have many edge cases. wdyt?

**Comment by p0mvn** [comment]
Good idea. However, while I agree that randomized tests would be useful and we could have a separate issue for it, I think that we still need some well-thought-out table-driven tests.

Table-driven unit tests help to ensure that we cover the edge cases that are expected by the programmer.

On the other hand, randomized tests help to detect unanticipated errors.

So, in my opinion, both are useful but for different reasons.

> python file generating all the random test values

I believe that @ValarDragon is working on introducing and improving a simulator that would help us with randomized tests across the repository without having to rely on external scripts. 

@catShaark  would you like to create a separate issue for this suggestion so that we continue discussing it there?

**Comment by catShaark** [comment]
> Good idea. However, while I agree that randomized tests would be useful and we could have a separate issue for it, I think that we still need some well-thought-out table-driven tests.
> 
> Table-driven unit tests help to ensure that we cover the edge cases that are expected by the programmer.
> 
> On the other hand, randomized tests help to detect unanticipated errors.
> 
> So, in my opinion, both are useful but for different reasons.
> 
> > python file generating all the random test values
> 
> I believe that @ValarDragon is working on introducing and improving a simulator that would help us with randomized tests across the repository without having to rely on external scripts.
> 
> @catShaark would you like to create a separate issue for this suggestion so that we continue discussing it there?

ohh, sure thing, lemme close this issue with good old table test first, then I'll put up a new issue like you said

**Comment by ValarDragon** [comment]
Randomized tests for these functions would look different to the simulator work! I wonder how they'd work with go fuzz

**Comment by catShaark** [comment]
I caught this unexpected behavior of `Pow`, It returns incorrect result ( off by so many digits ) for large base ( close to 2 ) and large exp. I guess we should have a new issue about this
```
		{
			// large base, large exp
			base:           sdk.MustNewDecFromStr("1.9999"),
			exp:            sdk.MustNewDecFromStr("123.456"),
			expectedResult: sdk.MustNewDecFromStr("14497004198753074745855987268357259398.755071824935534437"),
		},
```

**Comment by catShaark** [comment]
The problem partly comes from sdk.Dec Power, it's not correct for large exp input either, what should we do?

**Comment by catShaark** [comment]
> Was your decision to allow for error tolerance if it is within precision? 
I did not do anything to fix the problem. I think we have no choice but to simply accept it as a limitation of sdk.Dec. Well, If we really want to fix the problem, we have to fix sdk.Dec `Power` first which is a big deal so we should not include the fix into this PR anyway.
> Are you unblocked now?
I definitely don't know how to fix this precision problem.

**Comment by catShaark** [comment]
From playing with the test, I draw the conclusion that the larger the out put number, the worse precision we get
